torch/distributed/c10d_logger.py

```diff
@@ -0,0 +1,97 @@
+import logging
+import time
+from typing import List, Tuple
+
+from torch.distributed.logging_handlers import _log_handlers
+import torch.distributed as dist
+
+__all__: List[str] = ["exception_logger", "time_logger"]
+
+
+def _get_or_create_logger() -> logging.Logger:
+    logging_handler, log_handler_name = _get_logging_handler()
+    logger = logging.getLogger(f"c10d-{log_handler_name}")
+    logger.setLevel(logging.DEBUG)
+    formatter = logging.Formatter(
+        "%(asctime)s %(filename)s:%(lineno)s %(levelname)s p:%(processName)s t:%(threadName)s: %(message)s"
+    )
+    logging_handler.setFormatter(formatter)
+    logger.propagate = False
+    logger.addHandler(logging_handler)
+    return logger
+
+
+def _get_logging_handler(destination: str = "default") -> Tuple[logging.Handler, str]:
+    log_handler = _log_handlers[destination]
+    log_handler_name = type(log_handler).__name__
+    return (log_handler, log_handler_name)
+
+
+global _c10d_logger
+_c10d_logger = _get_or_create_logger()
+
+
+def exception_logger(func):
+    @functools.wraps(func)
+    def wrapper(*args, **kwargs):
+        try:
+            return func(*args, **kwargs)
+        except Exception as error:
+            if dist.is_initialized():
+                error_msg_dict = {
+                    "func_name": f"{func.__name__}",
+                    "args": f"{args}, {kwargs}",
+                    "backend": f"{dist.get_backend(kwargs.get('group'))}",
+                    "world_size": f"{dist.get_world_size(kwargs.get('group'))}",
+                    "global_rank": f"{dist.get_rank()}",
+                    "local_rank": f"{dist.get_rank(kwargs.get('group'))}",
+                    "error": f"{error}",
+                }
+            else:
+                error_msg_dict = {
+                    "func_name": f"{func.__name__}",
+                    "args": f"{args}, {kwargs}",
+                    "error": f"{error}",
+                }
+            _c10d_logger.debug(error_msg_dict)
+            raise
+    return wrapper
+
+
+def time_logger(func):
+    @functools.wraps(func)
+    def wrapper(*args, **kwargs):
+        t1 = time.time_ns()
+        func_return = func(*args, **kwargs)
+        t2 = time.time_ns()
+
+        if dist.is_initialized():
+            msg_dict = {
+                "func_name": f"{func.__name__}",
+                "args": f"{args}, {kwargs}",
+                "backend": f"{dist.get_backend(kwargs.get('group'))}",
+                "world_size": f"{dist.get_world_size(kwargs.get('group'))}",
+                "global_rank": f"{dist.get_rank()}",
+                "local_rank": f"{dist.get_rank(kwargs.get('group'))}",
+                "time_spent": f"{t2-t1}ns",
+            }
+        else:
+            msg_dict = {
+                "func_name": f"{func.__name__}",
+                "args": f"{args}, {kwargs}",
+                "time_spent": f"{t2-t1}",
+            }
+        _c10d_logger.debug(msg_dict)
+
+        return func_return
```

Comment:
Since there is a lot of overlap between `time_logger` and `exception_logger` I wonder if they should be combined into one logger that is configurable. WDYT? I guess it depends on if there is anything in the future we additionally want to surface.

Reply:
Ye. I was thinking about the same thing, but for exception_logger, we pretty much don't care about it being a synchronous call since the program is failing due to some errors anyway. 

For time_logger, it's not the case. If we merge both of them into one, there's some overhead(although tiny) for now if we combine them, so I decide to keep them separate for now to unblock the effort for logging time for the blocking calls. This is purely because we want to wrap some functions with @exception_logger and some functions with @time_logger only for the time being. 

Maybe for the following PRs, we can look into whether we can extract the repeating part in a helper function or merge the two loggers. WDYT?